### ISSUE TITLE
kvs: add FLUX_KVS_WAITCREATE flag for lookups

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1651]
As mentioned in #1556, the lightweight KVS watch proposed there returns an error if the watched key initially does not exist.  However, it's useful to be able to watch a key that may not yet exist, to avoid needing to synchronize out of band between producers and consumers of KVS data.

A FLUX_KVS_WAITCREATE flag could be used to modify FLUX_KVS_WATCH so that if the watched namespace and/or key have not yet been created, `kvs-watch.lookup` or `kvs-watch.getroot` do not immediately return an error.  Instead, the watch state is created, and the `kvs-watch` module subscribes to namespace creation events, and/or watches for mention of the watched key in a setroot, event.

Potentially we could support FLUX_KVS_WAITCREATE without FLUX_KVS_WATCH, in which case one would get back a single lookup result after creation and no subsequent updates.

**Comment by chu11** [comment]
I was pondering this and wondering if "WAIT_CREATE" isn't the right flag, or if it should be an additional flag on top of something else.

The use case pattern I was debating was the pattern where a user wants to know if the key does not exist before it is watched.  Right now the user would have to do a lookup before they watch:

```
f = flux_kvs_lookup (key)
if (flux_kvs_lookup_get (f, ...) < 0 && errno == ENOENT) {
    /* do whatever */
}
flux_kvs_lookup (key, FLUX_KVS_WATCH | FLUX_KVS_WAIT_CREATE)
```

I'm thinking, perhaps the flag we really want is something like "CONTINUE_ON_ENOENT", that even if a key does not exist, continue to watch it anyways?  This could work even if the key is deleted.

The "WAIT_CREATE" flag could presumably be a special case flag, saying "ignore the first ENOENT"?  Or perhaps its simply the caller's responsibility to ignore an ENOENT error if they set "CONTINUE_ON_ENOENT"?

**Comment by chu11** [comment]
hmmm, additional complication.  How would user be able to tell difference between ENOENT b/c key is missing vs ENOENT namespace is missing/removed.  Perhaps the latter should be ENOTSUP.  That's the errno generally used for illegal namespace.

**Comment by chu11** [comment]
after discussion at the meeting, it was determined that maintaining "true" backwards compatibility to the old KVS watch interface is probably not needed.  Many use cases in kvs tests can probably be dropped as unneeded/unsupported.  For example, watching a key that has been deleted and will come back in the future.

Instead the FLUX_KVS_WAITCREATE should be implemented just for its specific use case.

**Comment by chu11** [comment]
Subtlety I realized while working on this.

> kvs-watch module subscribes to namespace creation events

right now, there is no namespace creation event.  There are only setroot events, and setroot events only occur if a namespace is created AND something is written to it.

Should we have namespace create generate an event on its creation so it can be captured in a watch?

There's a part of me that is not inclined to do so, b/c a getroot is sort of meaningless on an empty namespace (i.e. one that is just created).  The reference pointing to an empty namespace can be pre-computed, so it's not actually a watch that's worthwhile.  We sort of care about the first "real" change.

So in other words, is WAITCREATE on a getroot for the namespace create?  Or for when the namespace is created & populated with something?  I think it should be the latter.

**Comment by chu11** [comment]
> So  in other words, is WAITCREATE on a getroot for the namespace create?   Or for when the namespace is created & populated with something?  I  think it should be the latter.

I'm going to answer my own question.  The language in the flux kvs manpage is clear, " If '-w' is specified, display the current root,", so we should create a namespace create event.

**Comment by garlick** [comment]
Oops, sorry for assuming that was already there.

I think I agree with you that adding the event is the right thing to do vis a vis semantics of WAITCREATE on a getroot or lookup.
